service/dps/storm/topologies/validation/src/main/java/eu/europeana/cloud/service/dps/storm/topologies/validation/topology/statistics/RecordStatisticsGenerator.java

```diff
@@ -0,0 +1,143 @@
+        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+        DocumentBuilder db = dbf.newDocumentBuilder();
+        InputSource is = new InputSource();
+        is.setCharacterStream(new StringReader(fileContent));
+        return db.parse(is);
+    }
+
+    private Element getRootElement(Document document) {
+        return document.getDocumentElement();
+    }
+
+    private void addRootToNodeList(Node root) {
+        String nodeXpath = getXPath(root);
+        String nodeValue = getFirstLevelTextContent(root);
+        String modelKey = getKey(nodeXpath, nodeValue);
+        nodeStatistics.put(modelKey, new NodeStatistics(nodeXpath, null, nodeValue, 1));
+    }
+
+    private void prepareNodesList(Node root) {
+        String parentXpath = getXPath(root);
+        NodeList childrenNodes = root.getChildNodes();
+        for (int i = 0; i < childrenNodes.getLength(); i++) {
+            Node node = childrenNodes.item(i);
+            if (node instanceof Element) {
+                String nodeXpath = getXPath(node);
+                String nodeValue = getFirstLevelTextContent(node);
+                String modelKey = getKey(nodeXpath, nodeValue);
+                NodeStatistics nodeModel = nodeStatistics.get(modelKey);
+                if (nodeModel == null) {
+                    nodeModel = new NodeStatistics(nodeXpath, parentXpath, nodeValue, 1);
+                } else {
+                    int currentOccurrence = nodeModel.getOccurrence();
+                    nodeModel.setOccurrence(++currentOccurrence);
+                }
+                assignAttributesToNode(nodeModel, node.getAttributes());
+                nodeStatistics.put(modelKey, nodeModel);
+                prepareNodesList(node);
+            }
+        }
+    }
+
+    private void assignAttributesToNode(NodeStatistics nodeModel, NamedNodeMap newAttributes) {
+        Set<AttributeStatistics> existedAttributes = nodeModel.getAttributes();
+        calculateAttributeStatistics(existedAttributes, newAttributes);
+        nodeModel.setAttributes(existedAttributes);
+    }
+
+    private void calculateAttributeStatistics(Set<AttributeStatistics> existedAttributes, NamedNodeMap attributes) {
+        for (int j = 0; j < attributes.getLength(); j++) {
+            Attr attr = (Attr) attributes.item(j);
+            String attrName = getAttributeXPath(attr);
+            String attrValue = attr.getNodeValue();
+            if (existedAttributes.isEmpty()) {
+                AttributeStatistics attributeModel = new AttributeStatistics(attrName, attrValue, 1);
+                existedAttributes.add(attributeModel);
+            } else {
+                AttributeStatistics attributeModel = new AttributeStatistics(attrName, attrValue, 1);
+                if (existedAttributes.contains(attributeModel)) {
+                    increaseOccurrence(existedAttributes, attributeModel);
+                }
+                existedAttributes.add(attributeModel);
```

Comment:
When the attribute statistics are present in the set increasing the occurrence is enough, no need to add the object again because it will not be added.